Add interface DigitalIO and AnalogIO

danielb987 opened the pull request:
A suggestion on how to add the interface DigitalIO.

pabender commented on java/test/jmri/LightTest.java:
```diff
@@ -31,6 +37,26 @@ public void testTransitionConstants() {
         Assert.assertTrue("TRANSITIONINGTOFULLOFF overlap", (Light.TRANSITIONINGTOFULLOFF & Light.TRANSITIONING) != 0);
     }
 
+    public void testLight() {
```
I think this test would be better as an addition to jmri.impelmentation.AbstractLightTestBase.  That way it gets run against all of the actual light implementations and not the MyLight class you added here.

danielb987 replied:
I agree. Fixed.

pabender commented on java/test/jmri/SensorTest.java:
```diff
@@ -0,0 +1,295 @@
+        Assert.assertTrue("Off and Inconsistent differ", (Sensor.OFF & Sensor.INCONSISTENT) == 0);
+    }
+
+    public void testSensor() throws JmriException {
```
I think this test would be better as an addition to jmri.impelmentation.AbstractSensorTestBase.  That way it gets run against all of the actual sensor implementations and not the MySensor class you added here.

danielb987 replied:
I agree. Fixed.

danielb987 commented:
I don't get jmri.jmrix.loconet.LnSensorTest to work. I try to fix it by adding a waitFor in AbstractSensorTestBase.testSensor()
```
public void testSensor() throws JmriException {
        t.setState(Sensor.ON);
        jmri.util.JUnitUtil.waitFor(()->{return t.getState() == Sensor.ON;}, "state = ON");
        Assert.assertTrue("Sensor is ON", t.getState() == Sensor.ON);
}
```
But I get the error `"state = ON" did not occur in time` when I run the test on my local computer. Any clue on how to fix it?

bobjacobsen commented:
There are a bunch of LnSensor tests that are disabled already, so disabling one more isn't such a bad thing.  The underlying cause for all of them is probably the setup not having all the correct jmrix.loconet pieces in place.  I'll open an Issue on that (I can't look at it now, as am trying to prioritize get the Internal Defaults code done)

danielb987 commented:
Travis fails on `jmri.jmrit.operations.automation.actions.GotoActionTest` method testAction at line 135.

It's the following code that fails on run2.join().

```
Thread run2 = JUnitUtil.getThreadByName("Run action item: " + automationItem2.getId());
try {
    run2.join();
} catch (InterruptedException e) {
    // do nothing
}
```

I have had that error many times on Travis. Should I add a check if run2 is null? Should I log a warning message if it's null? Or is it a critical error if run2 is null?

bobjacobsen commented:
I think I fixed a similar thing elsewhere. That problem was that the thread was completed already. Will try to see what I did to resolve...

danielb987 commented:
I have added an AnalogIO interface as a suggestion. I tried to keep AnalogIO aligned with DigitalIO.

danielb987 commented:
@bobjacobsen
> I think I fixed a similar thing elsewhere. That problem was that the thread was completed already. Will try to see what I did to resolve...

In GotoFailureActionTest there are tests that check if run and run2 is not null. I added the same tests to GotoActionTest.

danielb987 commented:
What are your thoughts about my interfaces DigitalIO and AnalogIO? Have I got it right or should it be done in some other way?

danielb987 commented on java/src/jmri/AnalogIO.java:
```diff
@@ -0,0 +1,80 @@
+    /**
+     * The minimum value of an analog value.
+     */
+    public static final int MIN_VALUE = Integer.MIN_VALUE;
```
I have changed AnalogIO to use float.

pabender commented:
@danielb987 thanks for turning that into a float.

The only other comment I have is on the limits.  Should restrict this value to be a signed percentage?

This is what I am thinking about ( as I look at using this on XBee hardware.... )

1) We don’t know what the actual measurement represents.  Even the hardware doesn’t necessarily know what the value represents ( We know that an n-bit ADC gives a value between 0 and 2^(n-1) and that 0 represents off and the maximum value typically represents the reference voltage, but we don’t have enough information to translate that into meaningful values )

2) since we don’t know what sensor is connected to the input, what other values are valid for us to compute in the General case?  
    A) We could send the raw ADC values, but then the user has to do interpretation of the data without knowing for certain what the reference value is.
    B) We could translate the data into percentages, because we should ( on the system level ) know what n is.
    C) perhaps there is specific hardware we can translate into a broader range, but the question is does the data on the network tell us anything about what that range is?

3)  Other parts of the code that already handle Analog values ( throttles and lights ) use percentages, and there is something to be said about consistency.

danielb987 commented:
Yes, you are right. The MIN and MAX constants are not useful when AnalogIO is changed to float (and maybe they won't before either). I have removed them now.

bobjacobsen commented:
I was thinking of the value as an engineering quantity, e.g. colts or RPM, rather than a raw quantity I.e. ADC/DAC counts.  That’s why a float seemed more appropriate. 

C.f. The Multimeter case which started this. 

As an engineering quantity, minimum and maximum are useful to consider. 

(Even if you stick with percentages, having minimum is useful: does is go down to 0 or to -100%?)

Percentages made sense for throttles when we had _no_ _idea_ what speed actually resulted from a particular setting, and one could loosely argue that a decode speed step is n% of some notional full power. (It’s not, but people think of it that way)

Some uses/implementations of AnanlogIO might still use percentages, I.e. Insteon Lights. But others might be able to provide something else, so we should preserve that. 

(Later, not now because we don’t have enough experience, we might want to add a getUnits() method for e.g assisting a in building displays)

danielb987 commented:
I have added the methods getMin() and getMax() methods to AnalogIO which allows the implementation to tell the minimum and maximum values that are supported.

One thing to consider is a method getResolution(). For example, a 8 bit ADC with would have a resolution on 1.0, a min value of 0.0 and a max value of 255.0.

mattharris commented:
@danielb987 Sticking my head above the parapet after a prolonged hiatus from JMRI dev work, but taking the different use cases for percentage versus discrete values could give us something like:

- `min=0.0`; `max=1.0` for percentage-based things (such as light intensity)
- `min=0.0`; `max!=1.0` for linear 'discrete' values (such as a voltmeter, ammeter, speedometer)

(The second example doesn't necessarily need to have `min=0.0`; it would be whatever is appropriate for the item being measured/output)

A `getResolution()` method would, indeed, be useful as you mention.

danielb987 commented:
Thanks. I have now added the method getResolution().

danielb987 commented:
I think this PR is ready to be merged.

bobjacobsen commented:
RE 'jmri.jmrit.operations.automation.actions.GotoActionTest' thread failures:  Checking for `null` turns this into a non-test, as it'll pass even if the underlying code-under-test never even ran due to some kind of regression.  But given the structure, it's probably the best that can be done now, and there's quite a bit of precedent:  Lots of our tests are really "Did it complete?" instead of "Did it do, in detail, what was expected?", and that's still better than no test at all.
